Changing the zoom gesture from shift + mousewheel to ctrl + mousewheel.

PhoeniXAbhisheK opened the pull request:
Fixes #1371

PhoeniXAbhisheK commented:
Do I need to add the snapshot files (generated during the tests in my system) in my commit too..???

julienw commented:
> @julienw @canaltinova Should I add the snapshot files (generated during the tests in my system) in my commit too..???
> That's what causing the CircleCI tests to fail

Yes please :)
The tests are using the snapshots to check whether the result we get is the same as the stored snapshots. So they need to be checked in.

Of course we'll review these changes as well: sometimes a snapshot changes but we didn't expect it to change. (I don't know if that's the case here, we'll have to look :) )

PhoeniXAbhisheK commented:
Hi @canaltinova @julienw 
As discussed, updated the snapshots, and added the Fixes #<issue_number> in the first comment.
All tests have also passed successfully 😸 😸 😸 😸 😸 
Please review.

gregtatum commented:
@julienw @canaltinova 

I'm running into issues with this change on my Mac:

1. The pinch and zoom gesture works, but it is painfully slow, so it's not very easy to use.
2. The pinch and zoom doesn't have the same inertia properties as the scrollwheel either.
3. I have the "zoom" accessibility feature turned on, and by default it uses the "ctrl" keybinding, so it overrides this change.  <img width="581" alt="image" src="https://user-images.githubusercontent.com/1588648/53518736-48f50280-3a97-11e9-9c61-286c1ab51e71.png">
4. Chrome uses the shift+mousewheel to change from a zoom, to a scroll, so it's got some prior art here.
5. This change is disruptive for our existing user's learned habits.

For point 1. I don't know how to fix it. I checked out the mouse event, and there is nothing to differentiate a pinch/zoom from a normal wheel event. Point 2 exacerbates this feeling of slowness.

To address points 3, 4, and 5, I think we should maintain the old keyboard shortcut so that ctrl or shift allows zooming. What do you think?

julienw commented:
Yeah, I think it's fine to keep both shortcuts :)

PhoeniXAbhisheK commented:
Hi @julienw , 
So now should I change the conditions and the names from 'ctrl' to 'ctrl or shift'..???

gregtatum commented:
@PhoeniXAbhisheK Yes, if you could make sure that we maintain both shortcuts, so it would work with both ctrl and shift. Thanks for the work on this!

PhoeniXAbhisheK commented:
@gregtatum sure, I'll get on to this 😸 😸 😸 😸 😸

PhoeniXAbhisheK commented:
Hi @gregtatum @julienw 
This is how I am trying to approch the changes.

For css,
The currrent rules are as such

```
.chartViewportCtrlScroll.hidden {
  opacity: 0;
  transform: translateY(30px);
}
```
These can be made to 
```
.chartViewportCtrlShiftScroll.hidden {
  opacity: 0;
  transform: translateY(30px);
}
```


For objects,

From this
```
type State = {|
  containerWidth: CssPixels,
  containerHeight: CssPixels,
  containerLeft: CssPixels,
  .
  .
  .
  isCtrlScrollHintVisible: boolean,
  isSizeSet: boolean,
|};
```

To this
```
type State = {|
  containerWidth: CssPixels,
  containerHeight: CssPixels,
  containerLeft: CssPixels,
  .
  .
  .
  isCtrlShiftScrollHintVisible: boolean,
  isSizeSet: boolean,
|};
```


For functions,

From this
```
showCtrlScrollingHint() {
        // Only show this message if we haven't ctrl zoomed yet.
        if (this.props.hasZoomedViaMousewheel) {
          return;
        }

        const scrollId = ++this.ctrlScrollId;
        if (!this.state.isCtrlScrollHintVisible) {
          this.setState({ isCtrlScrollHintVisible: true });
        }
        setTimeout(() => {
          if (scrollId === this.ctrlScrollId) {
            this.setState({ isCtrlScrollHintVisible: false });
          }
        }, 1000);
      }
```

To this
```
showCtrlShiftScrollingHint() {
        // Only show this message if we haven't ctrl zoomed or scroll zoomed yet.
        if (this.props.hasZoomedViaMousewheel) {
          return;
        }

        const scrollId = ++this.ctrlShiftScrollId;
        if (!this.state.isCtrlShiftScrollHintVisible) {
          this.setState({ isCtrlShiftScrollHintVisible: true });
        }
        setTimeout(() => {
          if (scrollId === this.ctrlShiftScrollId) {
            this.setState({ isCtrlShiftScrollHintVisible: false });
          }
        }, 1000);
      }
```

Is this the right approach to go forward towards making this change..???

PhoeniXAbhisheK commented:
Also, which of these popups are more intuitive..???


**Option 1**
![ctrlshiftscroll](https://user-images.githubusercontent.com/16290904/53622900-56d68f00-3c21-11e9-92c7-2df9c81e94fb.png)

**Option 2**
![ctrlscroll-shiftscroll](https://user-images.githubusercontent.com/16290904/53622918-61912400-3c21-11e9-80f0-50880a542aa1.png)

julienw commented:
Maybe we can simplify the class/variable names and just use `scrollHint` instead of mentioning the keys.

For the hint, maybe we can show the text `Ctrl / Shift` both inside _one_ button ?

PhoeniXAbhisheK commented:
Yes, that sounds better, will do that.

Is this hint looking right..???
![ctrl-shift-scroll](https://user-images.githubusercontent.com/16290904/53632961-38ca5800-3c3c-11e9-9c5f-48d56a355635.png)

julienw commented:
This looks good to me at least; @gregtatum do you have another feedback ?

PhoeniXAbhisheK commented:
Hi @julienw 
I have made the changes to files.
But for the `src/test/components/Viewport.test.js` file I am not sure how to proceed with a few changes.

For codes like below 👇 

```

          // Note that clientX will get rounded, so it won't be _exactly_ at the
          // center, which is why we do some approximations later.
          const viewport = scrollAndGetViewport({
            deltaY: -100,
            ctrlKey: true,
            clientX: BOUNDING_BOX_LEFT + innerComponentWidth * 0.5 + marginLeft,
          });

```
what do I change here..???

Should I remove the `ctrlKey: true,` part, or add another `shiftKey: true` to it..???
But if I add `shiftKey: true,` I am not sure if it would be the right way to test.
AFAIK, the test will then check for both Shift AND Ctrl key, which is not what we intend to do.

Please help.

julienw commented:
>  But for the `src/test/components/Viewport.test.js` file I am not sure how to proceed with a few changes

Ideally we should run the same test twice, but once with `shiftKey` and once with `ctrlKey`. Maybe it's good enough to copy/paste this test and change this part, or if we should extract functions etc. Let's do it wyth copy/paste first and look at the result.

.
> Also, I just realized that since we now accept both `ctrl+scroll` or `shift+scroll` for zoom, the code change I made also accepts `ctrl+shift+scroll` for zooming.
Is this acceptable..??? or should I change the code..???

Yeah I think this is acceptable.

PhoeniXAbhisheK commented:
Okk, so should I create a similar `It('test case', () => {})` block.
Or a whole new `describe` block..???

PhoeniXAbhisheK commented:
@julienw the changes are made.
Kindly review 😸

julienw commented:
Thanks! Just a quick note that I won't be able to look at it before monday now, but maybe somebody else from the team will. Have a nice week-end!

PhoeniXAbhisheK commented:
Sure,
Hence I submitted the patch today itself.
You too have a great weekend 😸 😸 😸 😸 😸

PhoeniXAbhisheK commented:
@julienw @canaltinova Kindly review the PR 😄 😄 😄 😄 😄

julienw commented:
Passing the review over to @gregtatum who knows this code well.

PhoeniXAbhisheK commented:
Hi @gregtatum, thanks for the feedback.
This is how the hint looks now.
Good..???

![Screenshot (93)](https://user-images.githubusercontent.com/16290904/54177482-6757f680-44b8-11e9-8bf2-b92502391ea2.png)

gregtatum commented on src/test/components/Viewport.test.js:
```diff
@@ -173,7 +181,33 @@ describe('Viewport', function() {
         //             ^
         //             Mouse position
-        it('zooms the preview selection equally when the mouse is centered', () => {
+        it('ctrl zooms the preview selection equally when the mouse is centered', () => {
```
So it looks like that test function is only being used on the first test, and the rest of the tests remain duplicated. I think it would be more maintainable here to continue this pattern throughout. If it's too verbose with naming a function every time, you could inline this tester function into a single test, or do a `for of` loop. Here is an example:

```diff
diff --git a/src/test/components/Viewport.test.js b/src/test/components/Viewport.test.js
index 0dd46aec..9f27d73d 100644
--- a/src/test/components/Viewport.test.js
+++ b/src/test/components/Viewport.test.js
@@ -151,6 +151,7 @@ describe('Viewport', function() {
   });
 
   describe('mousewheel zooming using shift + mousewheel', function() {
+    const modifierKeys = [{ ctrlKey: true }, { shiftKey: true }];
     // These series of tests are very particular to margin differences, so check
     // different margin combinations separately.
     const marginTests = [
@@ -174,29 +175,34 @@ describe('Viewport', function() {
         //             ^
         //             Mouse position
         it('zooms the preview selection equally when the mouse is centered', () => {
-          const { scrollAndGetViewport } = setup({
-            marginLeft,
-            marginRight,
-          });
-
-          // Note that clientX will get rounded, so it won't be _exactly_ at the
-          // center, which is why we do some approximations later.
-          const viewport = scrollAndGetViewport({
-            deltaY: -100,
-            shiftKey: true,
-            clientX: BOUNDING_BOX_LEFT + innerComponentWidth * 0.5 + marginLeft,
-          });
-
-          // Assert that this zooms in equally.
-          expect(viewport.viewportLeft).toBeGreaterThan(0);
-          expect(viewport.viewportRight).toBeLessThan(1);
-          expect(viewport.viewportLeft + viewport.viewportRight).toBeCloseTo(1);
-
-          // Only do an additional viewport top/bottom check here.
-          expect(viewport).toMatchObject({
-            viewportTop: 0,
-            viewportBottom: BOUNDING_BOX_HEIGHT,
-          });
+          for (const modifierKey of modifierKeys) {
+            const { scrollAndGetViewport } = setup({
+              marginLeft,
+              marginRight,
+            });
+
+            // Note that clientX will get rounded, so it won't be _exactly_ at the
+            // center, which is why we do some approximations later.
+            const viewport = scrollAndGetViewport({
+              deltaY: -100,
+              clientX:
+                BOUNDING_BOX_LEFT + innerComponentWidth * 0.5 + marginLeft,
+              ...modifierKey,
+            });
+
+            // Assert that this zooms in equally.
+            expect(viewport.viewportLeft).toBeGreaterThan(0);
+            expect(viewport.viewportRight).toBeLessThan(1);
+            expect(viewport.viewportLeft + viewport.viewportRight).toBeCloseTo(
+              1
+            );
+
+            // Only do an additional viewport top/bottom check here.
+            expect(viewport).toMatchObject({
+              viewportTop: 0,
+              viewportBottom: BOUNDING_BOX_HEIGHT,
+            });
+          }
         });
 
         // |        Viewport        |
```

PhoeniXAbhisheK commented:
@gregtatum Yes, I just wanted you to confirm that this is what you were expecting.
Now that you have told it's right, I'll continue this through the file 😄 😄 😄 😄 😄

gregtatum commented on src/test/components/Viewport.test.js:
```diff
@@ -74,8 +74,25 @@ describe('Viewport', function() {
     });
   });
 
+  function testScrollingHint(eventOptions) {
```
Woops, left this comment on. Please ignore it.
